Starport blog tutorial and polling app

fadeev opened the pull request:
Added:

* Starport blog tutorial
* Starport polling app tutorial
* ~CosmWasm smart contract mini-guide~

<img width="1482" alt="Screenshot 2020-07-23 at 13 52 47" src="https://user-images.githubusercontent.com/332151/88268388-cf912c00-cceb-11ea-885b-5b2aba5714ab.png">

marbar3778 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+
+## Prerequisites
+
+- A [recent version of Go](https://golang.org/doc/install) installed.
```
need to add node as a prerequisites

fadeev replied:
Ok

marbar3778 commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+### NPM
+
+```
+npm install -g @tendermint/starport
````
is this npm or brew or both?

fadeev replied:
It's either. Yeah, I think link to a readme makes sense.

marbar3778 commented on starport-blog/02-cli-list-posts.md:
````diff
@@ -0,0 +1,141 @@
+}
+```
+
+`GetCmdListPosts` runs an ABCI query to fetch the data, unmarshals it back form binary to JSON and returns it to the console. ABCI is an interface between your app and Tendermint (a program responsible for replicating the state across machines). ABCI queries look like paths on a hierarchical filesystem. In our case, the query is `custom/blog/list-posts`. Before we continue, we need to define `Que ...
````
link to abci docs, maybe?

fadeev replied:
Will do.

marbar3778 commented on starport-smart-contracts/index.md:
```diff
@@ -0,0 +1,213 @@
+# Smart contracts with CosmWasm
```
could we do this in a separate PR? I think we could coordinate something with the cosmwasm team instead of having our own tutorial.

marbar3778 commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+}
+```
+
+## x/blog/alias.go
````
alias.go is no longer needed and not in the module spec. we can remove it from starport?

fadeev replied:
Is there a doc I can read about this? Because alias.go is still in the Nameservice, so I figured it's still the best practice.

marbar3778 commented:
It seems there is a lot of things being done behind the scene. The app.go file where everything is pulled together wasn't touched on and its the most important part of putting a app together. Im not really sure how the chain functions other than creating a blog. I think this is an important part of this education part. What do others think?

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+
+Alternatively, Starport can be built [from source](http://github.com/tendermint/starport).
+
+After `starport` is installed, use it to create the initial app structure:
```
```suggestion
After `starport` is installed, use it to create the initial app structure inside a directory named `blog`:
```

okwme replied:
can this be customized?

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+)
+```
+
+This file already contains `func GetTxCmd` which defines custom `blogcli` commands. We will add the custom `create-post` command to our `blogcli` by first adding `GetCmdCreatePost`.
````
```suggestion
This file already contains `func GetTxCmd` which defines custom `blogcli` commands. We will add the custom `create-post` command to our `blogcli` by first adding `GetCmdCreatePost` to `blogTxCmd`.
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+}
+```
+
+The function above defines what happens when you run `create-post` subcommand. `create-post` takes one argument `[title]`, creates a message `NewMsgCreatePost` (with title as `args[0]`) and broadcasts this message.
````
```suggestion
The function above defines what happens when you run the `create-post` subcommand. `create-post` takes one argument `[title]`, creates a message `NewMsgCreatePost` (with title as `args[0]`) and broadcasts this message.
```

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+The function above defines what happens when you run `create-post` subcommand. `create-post` takes one argument `[title]`, creates a message `NewMsgCreatePost` (with title as `args[0]`) and broadcasts this message.
+This is a common pattern in the SDK: users make changes to the store by broadcasting messages. Both CLI commands and HTTP requests create messages that can be broadcasted in order for state transition to occur.
+
+Let’s define `NewMsgCreatePost`.
```
```suggestion
Let’s define `NewMsgCreatePost` by creating a new file at `x/blog/types/MsgCreatePost.go`.
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+}
+```
+
+`NewMsgCreatePost` is a constructor function that creates `MsgCreatePost` message. The following five functions have to be defined to implement `Msg` interface. They allow you to perform validation that doesn’t require access to store (like checking for empty values), etc.
````
```suggestion
`NewMsgCreatePost` is a constructor function that creates the `MsgCreatePost` message. The following five functions have to be defined to implement the `Msg` interface. They allow you to perform validation that doesn’t require access to the store (like checking for empty values), etc.
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+}
+```
+
+Going back to `GetCmdCreatePost`, you created a new `MsgCreatePost` message and broadcasted it with `GenerateOrBroadcastMsgs`.
````
```suggestion
Going back to `GetCmdCreatePost` in `x/blog/client/cli/tx.go`, you'll see `MsgCreatePost` being created and broadcast with `GenerateOrBroadcastMsgs`.
```

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+
+Going back to `GetCmdCreatePost`, you created a new `MsgCreatePost` message and broadcasted it with `GenerateOrBroadcastMsgs`.
+
+After being broadcasted, the messages are processed by handlers.
```
```suggestion
After being broadcast, the messages are processed by an important part of the SDK, called **handlers**.
```

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+After being broadcasted, the messages are processed by handlers.
+
+## x/blog/handler.go
+
```
```suggestion
Begin by importing your new blog types that we created:
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+
+```go
+    switch msg := msg.(type) {
+    case MsgCreatePost:
````
```suggestion
    case types.MsgCreatePost:
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+Now let’s define `handleMsgCreatePost`:
+
+```go
+func handleMsgCreatePost(ctx sdk.Context, k Keeper, msg MsgCreatePost) (*sdk.Result, error) {
````
```suggestion
func handleMsgCreatePost(ctx sdk.Context, k Keeper, msg types.MsgCreatePost) (*sdk.Result, error) {
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+}
+```
+
+In this handler you create a `Post` object (post type was defined in the very first step). You populate the post object with creator and title from the message (`msg.Creator` and `msg.Title`) and generate a unique ID with `uuid.New().String()`. Make sure to add `"github.com/google/uuid"` to the import at the top of this file.
````
is `uuid.New()` deterministic?

okwme replied:
```suggestion
In this handler you create a `Post` object (post type was defined in the very first step). You populate the post object with creator and title from the message (`msg.Creator` and `msg.Title`) and use the unique ID that was generated in `tx.go` with `NewMsgCreatePost()` using `uuid.New().String()`.
```

fadeev replied:
> is uuid.New() deterministic?

It is not, but it is used during message creation, so I didn't see a problem in that.

okwme replied:
if it's created in message generation you're right there's no problem, but is there also a check happening to confirm uniqueness?

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+
+In this handler you create a `Post` object (post type was defined in the very first step). You populate the post object with creator and title from the message (`msg.Creator` and `msg.Title`) and generate a unique ID with `uuid.New().String()`. Make sure to add `"github.com/google/uuid"` to the import at the top of this file.
+
+After creating a post object with creator, ID and title, message handler calls `k.CreatePost(ctx, post)`. “k” stands for Keeper, an abstraction used by the SDK that writes data to the store. Let’s define `CreatePost` keeper function.
```
```suggestion
After creating a post object with creator, ID and title, the message handler calls `k.CreatePost(ctx, post)`. “k” stands for Keeper, an abstraction used by the SDK that writes data to the store. Let’s define the `CreatePost` keeper function.
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+}
+```
+
+`CreatePost` creates a key by concatenating a post prefix with an ID. If you look back at how our store looks, you’ll notice keys have prefixes `post-0bae9f7d-20f8-4b51-9d5c-af9103177d66` in this case `post-` . The reason for this is you have one store, but you might want to keep different types of objects in it, like posts and users. Prefixing keys with `post-` and `user-` allows you to share on ...
````
```suggestion
`CreatePost` creates a key by concatenating a post prefix with an ID. If you look back at how our store looks, you’ll notice keys have prefixes, which is why `post-0bae9f7d-20f8-4b51-9d5c-af9103177d66` contained the prefix `post-` . The reason for this is you have one store, but you might want to keep different types of objects in it, like posts and users. Prefixing keys with `post-` and `user-` allows you to share one storage space between different types of objects.
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+Run the following command to create a post:
+
+```sh
+blogcli tx blog create-post 'This is a post!" --from=user1
````
```suggestion
blogcli tx blog create-post "This is a post!" --from=user1
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+blogcli tx blog create-post 'This is a post!" --from=user1
+```
+
+“This is a post!” is a title for our post and `--from=user1` tells the program who is creating this post. `user1` is a label for your pair of keys used to sign the transaction, created by initialization script previously. Keys are stored in `~/.blogcli`.
````
```suggestion
“This is a post!” is a title for our post and `--from=user1` tells the program who is creating this post. `user1` is a label for your pair of keys used to sign the transaction, created by the initialization script located within the `/Makefile` previously. Keys are stored in `~/.blogcli`.
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+http://localhost:26657/tx?hash=0xCA1491B39384A4F29E568F62B156E0F2D0601507EF499CE1B8F3930BAFE7F03C
+```
+
+Congratulations! You have just created and launched your custom blockhain and sent the first transaction.
````
I got this error:
```
error converting http params to arguments: invalid character 'D' looking for beginning of value
```

okwme replied:
i see was missing the `0x`

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+
+After running the command and confirming it, you will see an object with “txhash” property with a value like `CA1491B39384A4F29E568F62B156E0F2D0601507EF499CE1B8F3930BAFE7F03C`.
+
+To verify that the transaction has been processed, open a browser and visit the following URL (make sure to replace `CA14...` with the value of your txhash):
```
```suggestion
To verify that the transaction has been processed, open a browser and visit the following URL (make sure to replace `CA14...` with the value of your txhash but make sure to keep the `0x` prefix):
```

okwme commented on starport-blog/01-index.md:
````diff
@@ -0,0 +1,350 @@
+http://localhost:26657/tx?hash=0xCA1491B39384A4F29E568F62B156E0F2D0601507EF499CE1B8F3930BAFE7F03C
+```
+
+Congratulations! You have just created and launched your custom blockhain and sent the first transaction.
````
```suggestion
Congratulations! You have just created and launched your custom blockchain and sent the first transaction 🎉
```

okwme commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,350 @@
+
+Congratulations! You have just created and launched your custom blockhain and sent the first transaction.
+
+## Errors
```
What is this section? a FAQ? Why is it here?

fadeev replied:
It's a list of error that developers might encounter. If they do, finding answers online may prove challenging.

okwme commented on starport-blog/02-cli-list-posts.md:
````diff
@@ -0,0 +1,141 @@
+)
+```
+
+Function GetQueryCmd is used for creating a list of `query` subcommands, it should already be defined. Edit the function to add `GetCmdListPosts` as a subcommand:
````
```suggestion
Function `GetQueryCmd` is used for creating a list of `query` subcommands, it should already be defined. Edit the function to add `GetCmdListPosts` as a subcommand:
```

okwme commented on starport-blog/02-cli-list-posts.md:
````diff
@@ -0,0 +1,141 @@
+}
+```
+
+This function uses an prefix iterator to loop through all the keys with a given prefix (in our case `PostPrefix` is `"post-"`). We’re getting values by key with `store.Get` and appending them to postList. Finally, we unmarshal bytes back to JSON and return the result to the console.
````
```suggestion
This function uses a prefix iterator to loop through all the keys with a given prefix (in our case `PostPrefix` is `"post-"`). We’re getting values by key with `store.Get` and appending them to postList. Finally, we unmarshal bytes back to JSON and return the result to the console.
```

okwme commented on starport-blog/02-cli-list-posts.md:
```diff
@@ -0,0 +1,141 @@
+
+This function uses an prefix iterator to loop through all the keys with a given prefix (in our case `PostPrefix` is `"post-"`). We’re getting values by key with `store.Get` and appending them to postList. Finally, we unmarshal bytes back to JSON and return the result to the console.
+
+Now let’s see how it works. Run the following command to recompile your app, clear the data and relaunch the chain:
```
checking back to my console i have this error:
```
2020/07/24 16:19:12 Error in building the application. Please, check ./Makefile
```
Which is good in that the console has stopped so i can re-run `starport serve` but it seems that shouldn't be the case. Should the original `starport serve` have stopped already or should there be an instruction to stop it? Or was this supposed to be part of a hot reload?

fadeev replied:
You probably saved a file, hot-reload tried rebuilding and failed and it ended up with an error. We need to rewrite `starport serve` so that a build error doesn't exit Starport, but rather makes it wait for further changes (sort of like on the frontend).

okwme commented on starport-blog/03-http-list-posts.md:
```diff
@@ -0,0 +1,85 @@
+# List posts with HTTP
+
+At some point you may want to create a front-end to your application and interact with it by sending HTTP requests. In part 2 we've seen that sending requests to `/abci_query` is one way of getting data from your app. This works well for testing and certain specific usecases, but for a production application you may want to create a clean API for yourself and other users. This API should probably ...
```
```suggestion
At some point you may want to create a front-end to your application and interact with it by sending HTTP requests. In part 2 we've seen that sending requests to `/abci_query` is one way of getting data from your app. This works well for testing and certain specific usecases, but for a production application you may want to create a clean API for yourself and other users. This API should probably have short endpoints (without implementation details, like `abci`) and respond differently to HTTP verbs like GET and POST. You may also want to have some request processing logic that handles basic checks and data transformation. Luckily, Cosmos SDK provides just that.
```

okwme commented on starport-blog/03-http-list-posts.md:
```diff
@@ -0,0 +1,85 @@
+import (
+  // Existing imports...
+  "fmt"
+	"net/http"
```
```suggestion
  "net/http"
```

okwme commented on starport-blog/03-http-list-posts.md:
````diff
@@ -0,0 +1,85 @@
+Add the following line to register our first route:
+
+```go
+	r.HandleFunc("/blog/posts", listPostsHandler(cliCtx, "blog")).Methods("GET")
````
Why is `	r.HandleFunc("/faucet", faucetHandler(cliCtx)).Methods("POST")` already included in this file?

fadeev replied:
The template used to have a faucet, this should be removed.

https://github.com/tendermint/starport/issues/55

okwme commented on starport-blog/03-http-list-posts.md:
````diff
@@ -0,0 +1,85 @@
+To build your app and launch servers, run:
+
+```
+starport serve
````
the previous `serve` command hadn't crashed this time. On save it looked like it refreshed, was this a hot reload? Should i cancel the CLI session and restart it anyway?

fadeev replied:
Yes, it's hot reload. If `serve` is running, the user shouldn't run it again. I'll add "if `serve` is not running, run..."

okwme commented on starport-blog/03-http-list-posts.md:
````diff
@@ -0,0 +1,85 @@
+To launch our HTTP server run the following command in a different terminal window:
+
+```
+blogcli rest-server
````
I got this error:
```
ERROR: failed to listen on localhost:1317: listen tcp 127.0.0.1:1317: bind: address already in use
```

okwme replied:
I ran `killall -9 blogcli` and tried it again and it worked

fadeev replied:
Yeah, evidently the process wasn't killed.

okwme commented on starport-blog/04-http-create-post.md:
```diff
@@ -0,0 +1,282 @@
+import (
+  // Existing imports...
+  "github.com/example/blog/x/blog/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
```
```suggestion
  "github.com/cosmos/cosmos-sdk/x/auth/client/utils"
```

okwme replied:
wy is this here actually? i don't see it being used.

fadeev replied:
You're right, I'll remove it.

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+	r.HandleFunc("/blog/posts", createPostHandler(cliCtx)).Methods("POST")
+```
+
+Now let's create `createPostHandler`.
````
```suggestion
Now let's create `createPostHandler` inside a new file you should create at `x/blog/client/rest/tx.go`.
```

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+}
+```
+
+`createPostHandler` first parses request parameters, performs basic validations, converts `Creator` field from string into SDK account address type, creates `MsgCreatePost` message.
````
```suggestion
`createPostHandler` first parses request parameters, performs basic validations, converts `Creator` field from a string into an SDK account address type then finally creates `MsgCreatePost` message.
```

okwme commented on starport-blog/04-http-create-post.md:
```diff
@@ -0,0 +1,282 @@
+
+## Setting up the client-side project
+
+Inside an empty directory create `package.json`:
```
```suggestion
Inside a new empty directory create `package.json`:
```

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+}
+```
+
+Responses from an HTTP server provided by `blogcli rest-server` do not provide `Access-Control-Allow-Origin` headers. This prevents client-side application running from different domains/ports from making successful requests due to browser's CORS policy. In production that can be managed by a web server like Nginx providing the right headers. In development we'll be using `local-cors-proxy` (`lcp ...
````
```suggestion
Responses from an HTTP server provided by `blogcli rest-server` do not provide `Access-Control-Allow-Origin` headers. This prevents client-side application running from different domains/ports from making successful requests due to most browsers' CORS policies. In production that can be managed by a web server like Nginx providing the right headers. In development we'll be using `local-cors-proxy` (`lcp`) to act as a proxy between our front-end and server.
```

okwme commented on starport-blog/04-http-create-post.md:
```diff
@@ -0,0 +1,282 @@
+
+Responses from an HTTP server provided by `blogcli rest-server` do not provide `Access-Control-Allow-Origin` headers. This prevents client-side application running from different domains/ports from making successful requests due to browser's CORS policy. In production that can be managed by a web server like Nginx providing the right headers. In development we'll be using `local-cors-proxy` (`lcp ...
+
+`preserve` runs first and launches `lcp` in background. `serve` launches a development server and `postserve` kills `lcp` process after you shutdown the server.
```
```suggestion
The commands under `scripts` work as follows: `preserve` runs first and launches `lcp` in background. `serve` launches a development server and `postserve` kills `lcp` process after you shutdown the server.
```

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+  createBroadcastTx,
+} from "@tendermint/sig";
+```
+
````
why a gap here?

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+const wallet = createWalletFromMnemonic(mnemonic);
+```
+
+If you use the same mnemonic (think of it as a password), you will get the same wallet address: in out case `wallet.address` is `cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh`. If you want to generate unique mnemonics for your users, you can use `bip39` package.
````
link bip39?

okwme commented on starport-blog/04-http-create-post.md:
```diff
@@ -0,0 +1,282 @@
+
+`fetchPosts` makes an HTTP GET request to `/blog/posts` and inserts the posts into the `<div id="posts">` container.
+
+Next, we need to define `createPost` function. It will take the value from the text input, fetch required account parameters from the server, fetch an unsigned transaction from the server, sign it using our private key from the wallet, broadcast it back to the server and fetch the list of posts again.
```
```suggestion
Next, we need to define the `createPost` function. It will take the value from the text input, fetch the required account parameters from the server, fetch an unsigned transaction from the server, sign it using our private key from the wallet, broadcast it back to the server and fetch the list of posts again.
```

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+Transfer some tokens from this account to the new one:
+
+```sh
+blogcli tx send cosmos1wt47yve6l29yjtxtsajhltr2vqhf7mpw5n6fx6 cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh 10foo --from=user1
````
```suggestion
blogcli tx send $(blogcli keys show user1 -a) cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh 10foo --from=user1
```

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+blogcli tx send cosmos1wt47yve6l29yjtxtsajhltr2vqhf7mpw5n6fx6 cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh 10foo --from=user1
+```
+
+Notice that the sender address `cosmos1wt47yve6l29yjtxtsajhltr2vqhf7mpw5n6fx6` is the one we got from running `blogcli keys show me` and the receiver account address `cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh` is the one we have generated from the mnemonic in browser.
````
```suggestion
Notice that the sender address can be queried automatically using the sub-command `$(blogcli keys show user1 -a)` with the flag `-a` to show just the address and the receiver account address `cosmos152gzu3vzf7g9tu46vszgpac24lwr48vc8k8kkh` is the one we have generated from the mnemonic in the browser.
```

okwme commented on starport-blog/04-http-create-post.md:
```diff
@@ -0,0 +1,282 @@
+
+In this guide we're activating accounts manually, but in production apps you might want to do it as part of a signing up process.
+
+Now that you that we've created the app and set up our account, let's run it!
```
```suggestion
Now that we've created the app and set up our account, let's run it!
```

okwme commented on starport-blog/04-http-create-post.md:
````diff
@@ -0,0 +1,282 @@
+npm run serve
+```
+
+Open `http://localhost:1234/` and try creating new posts. New posts should show up in the list after 2-4 seconds after submission. Notice that posts persist after you refresh the page.
````
I get this error in the browser console:
```
xhr.js:178 POST http://localhost:8010/proxy/blog/posts 405 (Method Not Allowed)
```

fadeev replied:
Seems like proxy didn't work. I've created a repo with branches for each step. Checked multiple times, couldn't reproduce this error.

https://github.com/fadeev/tutorial-blog/tree/04-http-create-posts

fadeev replied:
There was one more error: `foo` instead of `token`. Starport used to init with `foo` (like Scavenge), but we have replaced it with `token`.

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+We will be creating a simple poll application, in which a user can sign in, create polls, cast votes and see voting results. Creating a poll will cost 200 tokens, voting is free, and both actions will be available only for signed in users.
+
+For this tutorial we will be using [Starport](https://github.com/tendermint/starport), an easy to use tool for building blockchains. Install Starport and run the following command to create a voter project:
```
link to install?

okwme commented on starport-polling-app/index.md:
````diff
@@ -0,0 +1,277 @@
+starport type poll title options
+```
+
+This command generated code that handles creation of `poll` items. If we now run `starport serve` and visit [http://localhost:8080](http://localhost:8080) we will see a form for creating polls. It may take a short while to rebuild the app, so give it a couple of seconds.
````
```suggestion
This command generated code that handles the creation of `poll` items. If we now run `starport serve` and visit [http://localhost:8080](http://localhost:8080) we will see a form for creating polls. It may take a short while to rebuild the app, so give it a couple of seconds.
```

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+This command generated code that handles creation of `poll` items. If we now run `starport serve` and visit [http://localhost:8080](http://localhost:8080) we will see a form for creating polls. It may take a short while to rebuild the app, so give it a couple of seconds.
+
+![Application screenshot](./2.png)
```
I just see the sign-in form

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+![Application screenshot](./2.png)
+
+Sign in with one of the passwords printed in the console and try creating a poll. You should see a new object created and displayed above the form. You have successfully created an object and stored it on the blockchain!
```
```suggestion
Sign in with one of the passwords printed in the console after running `starport serve` and try creating a poll. You should see a new object created and displayed above the form. You have successfully created an object and stored it on the blockchain!
```

okwme replied:
still don't have anything but the login screen

okwme replied:
actually it's showing the blog login screen. I've tried:
```
killall -9 starport
killall -9 blogd
killall -9 blogcli
```
but it's still runnning

okwme replied:
ok `killall -9 node` seemed to work although i saw two `serve` processes running with `htop`

fadeev replied:
You may have started two `starport serve` proccesses and the second one started the frontend on `:8081` instead. So you were seeing the sign in form, because it was a different frontend process running. As I've mentioned, we should rewrite `serve`, so that it 1. handles errors better. 2. kills and restarts proccesses more reliably (e.g. prevent frontend from running on a different port, etc.)

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+### `x/voter/types/TypePoll.go`
+
+This file contains definition of the `Poll` type. We can see that a poll has two fields (creator and ID), which will be created automatically, and two fields (title and options) defined by us. Since we want `Options` to be a list of string, **replace `string` with `[]string`**
```
```suggestion
This file contains definition of the `Poll` type. We can see that a poll has two fields (creator and ID), which will be created automatically, and two fields (title and options) defined by us. Since we want `Options` to be a list of strings, **replace `string` with `[]string`**
```

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+### `frontend/src/view/Index.vue`
+
+Since we don't need the default form component replace `<type-list />` with a new component `<poll-form />`
```
```suggestion
Since we don't need the default form component replace `<type-list />` inside of `frontent/src/views/Index.vue` with a new component `<poll-form />` that will be created in a new file at `frontend/src/components/PollForm.vue`.
```

okwme commented on starport-polling-app/index.md:
````diff
@@ -0,0 +1,277 @@
+</script>
+```
+
+Poll form component has an input for title and a list of inputs for options. Clicking "Add option" button adds an empty input and clicking "Create poll" sends a creates and broadcasts a transaction with a "create poll" message.
````
```suggestion
Poll form component has an input for title and a list of inputs for options. Clicking "Add option" button adds an empty input and clicking "Create poll" sends, creates and broadcasts a transaction with a "create poll" message.
```

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+### `frontend/src/views/Index.vue`
+
+Add `<poll-list />` after the poll form component.
```
```suggestion
Add `<poll-list />` into the `frontend/src/view/Index.vue` file after the poll form component.
```

okwme commented on starport-polling-app/index.md:
````diff
@@ -0,0 +1,277 @@
+</script>
+```
+
+Poll list component lists for every poll lists all options as buttons. Selecting an option triggers a `submit` method that broadcasts transaction with "create vote" message and fetches data back from our application.
````
```suggestion
The `PollList` component lists for every poll, all the options for that poll, as buttons. Selecting an option triggers a `submit` method that broadcasts a transaction with a "create vote" message and fetches data back from our application.
```

okwme replied:
That means you cast a vote just by clicking, so choose wisely!

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+To fix this issue we first have to understand how data is stored in our application.
+
+We can think of our data storage as a lexiographically order key value store. You can loop through the entries, filter by key prefix, add, update and delete entries. It is easier to visualize the store as a JSON:
```
```suggestion
We can think of our data storage as a lexicographically ordered key value store. You can loop through the entries, filter by key prefix, add, update and delete entries. It is easier to visualize the store as JSON:
```

okwme commented on starport-polling-app/index.md:
````diff
@@ -0,0 +1,277 @@
+key := []byte(types.VotePrefix + vote.PollID + "-" + string(vote.Creator))
+```
+
+Restart the application and try voting multiple times on a single poll.
````
```suggestion
Restart the application and try voting multiple times on a single poll, you'll see you can vote as many times as you want but only your most recent vote is counted.
```

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+Let's make it so that creating a poll costs 200 tokens.
+
+This feature is very easy to add. We already require users to have accounts registered, and each user has tokens on balance. The only thing we need to do is to send coins from user's account to a module account before we create a poll. `x/voter/handlerMsgCreatePoll.go`:
```
```suggestion
This feature is very easy to add. We already require users to have accounts registered, and each user has tokens on balance. The only thing we need to do is to send coins from user's account to a module account before we create a poll. 

## `x/voter/handlerMsgCreatePoll.go`:
```

okwme commented on starport-polling-app/index.md:
```diff
@@ -0,0 +1,277 @@
+
+Add the code above before `k.CreatePoll(ctx, poll)`. This way, if a user does not have enough tokens, the application will raise an error and will not proceed to creating a poll. Make sure to add `"[github.com/tendermint/tendermint/crypto](http://github.com/tendermint/tendermint/crypto)"` to the import statement (if your text editor didn't do that for you).
+
+Now, restart the app and try creating several polls to see this affects your token balance.
```
```suggestion
Now, restart the app and try creating several polls to see how this affects your token balance.
```

okwme commented on starport-smart-contracts/index.md:
```diff
@@ -0,0 +1,213 @@
+# Smart contracts with CosmWasm
+
+Cosmos application are often written in Go using Cosmos SDK. The code gets compiled before the chain is launched and effectively the app becomes an intergral part of the chain. If you ever need to update the code of the application, the chain has to go though a process of upgrading.
```
```suggestion
Cosmos application are often written in Go using the Cosmos SDK. The code gets compiled before the chain is launched and effectively the app becomes an intergral part of the chain. If you ever need to update the code of the application, the chain has to go though a process of upgrading.
```

fadeev commented:
I've added a short overview of files and directories generated by the scaffolding and a quick intro to the contents of the `frontend` directory.

I've also removed CosmWasm mini-guide as per Marko's request, but I'll create a new PR with it, because I think it's a very short and useful read on how to enable contracts—something that I'm sure many will find interesting (even if it mirrors _some_ of the content from CosmWasm docs).

Thank you, @okwme and @marbar3778, for providing so much feedback, I've tried to incorporate all of your suggestions.

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+
+## Overview
+
+Let's take a quick look at what Starport has generated for us. `app/app.go` file imports and configures SDK modules and creates a constructor for our application that extends a [basic SDK application](https://docs.cosmos.network/master/core/baseapp.html) among other things. This app will use only a couple standard modules bundled with Cosmos SDK (including `auth` for dealing with accounts and `ba ...
```
Might be worth linking to this doc https://docs.cosmos.network/master/basics/app-anatomy.html

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+
+Let's take a quick look at what Starport has generated for us. `app/app.go` file imports and configures SDK modules and creates a constructor for our application that extends a [basic SDK application](https://docs.cosmos.network/master/core/baseapp.html) among other things. This app will use only a couple standard modules bundled with Cosmos SDK (including `auth` for dealing with accounts and `ba ...
+
+In `cmd` directory we have source files of two programs for interacting with our application: `blogd` can configure genesis and launch the app and `blogcli` contains commands to send transactions and start an HTTP API server.
```
`blogd` starts a full-node for your blockchian

`blogcli`  enables you to query the full-node, either to update the state by sending a transaction or to read it via a query

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+
+In `cmd` directory we have source files of two programs for interacting with our application: `blogd` can configure genesis and launch the app and `blogcli` contains commands to send transactions and start an HTTP API server.
+
+This blog app will store data in a persistent key-value store. Similarly to most key-value stores, you can retrieve, delete, update, and loop through keys to obtain the values you are interested in.
```
link to https://docs.cosmos.network/master/core/store.html

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+Let’s define `NewMsgCreatePost` in a new file you should create as `x/blog/types/MsgCreatePost.go`.
+
+## x/blog/types/MsgCreatePost.go
+
```
It's a bit weird to start a section with a code block. A short intro sentence would be nice

gamarin2 replied:
maybe just take the previous sentence and put it there

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+
+Going back to `GetCmdCreatePost` in `x/blog/client/cli/tx.go`, you'll see `MsgCreatePost` being created and broadcast with `GenerateOrBroadcastMsgs`.
+
+After being broadcast, the messages are processed by an important part of the SDK, called **handlers**.
```
part of the "application"

gamarin2 replied:
+ link to https://docs.cosmos.network/master/building-modules/handler.html

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+
+In this handler you create a `Post` object (post type was defined in the very first step). You populate the post object with creator and title from the message (`msg.Creator` and `msg.Title`) and use the unique ID that was generated in `tx.go` with `NewMsgCreatePost()` using `uuid.New().String()`.
+
+After creating a post object with creator, ID and title, the message handler calls `k.CreatePost(ctx, post)`. “k” stands for Keeper, an abstraction used by the SDK that writes data to the store. Let’s define the `CreatePost` keeper function.
```
link to https://docs.cosmos.network/master/building-modules/keeper.html

gamarin2 commented on starport-blog/01-index.md:
```diff
@@ -0,0 +1,345 @@
+After creating a post object with creator, ID and title, the message handler calls `k.CreatePost(ctx, post)`. “k” stands for Keeper, an abstraction used by the SDK that writes data to the store. Let’s define the `CreatePost` keeper function.
+
+## x/blog/keeper/keeper.go
+
```
Same as above, it's not good practice to start a section with a code block

gamarin2 commented on starport-blog/02-cli-list-posts.md:
```diff
@@ -0,0 +1,141 @@
+To list created posts we will be using `blogcli query blog list-posts` command. list-posts subcommand hasn’t been defined yet, so let’s do it now. Query commands from the CLI are handled by `query.go`.
+
+## x/blog/client/cli/query.go
+
```
Same

fadeev replied:
Done!

fadeev commented:
Added descriptions after headers and a bunch of links to docs (as per @gamarin2 suggestion).

fadeev commented:
@okwme @marbar3778 @gamarin2 Do we need to make any other changes to merge this PR?
